File: Exam_Review/Data_Structures/Linked_List.c
#include "Linked_List.h"
#include <stdio.h>
#include <stdlib.h>

// Linked List (LL)

// typedef struct Node {
//     int elem;
//     struct Node *next;
// } Node;

// typedef struct LinkedList {
//     Node *head;
// } LL;

Node* create_node(int elem) {
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->elem = elem;
    new_node->next = NULL;
    return new_node;
}

void create_list(LL **list) {
    *list = (LL *)malloc(sizeof(LL));
    (*list)->head = NULL;
}

void insert(LL *list, int elem) {
    Node *new_node = create_node(elem);

    if (list->head == NULL) {
        list->head = new_node;
    } else {
        Node *cur_node = list->head;
        while (cur_node->next != NULL) {
            cur_node = cur_node->next;
        }
        cur_node->next = new_node;
    }
}

File: Exam_Review/Data_Structures/Linked_List.py
class Node:
    def __init__(self, value):
        self.val = value
        self.next = None

class LL:
    def __init__(self):
        self.head = None

    def append(self, val):
        new = Node(val)
        if not self.head:
            self.head = new
            return
        
        cur = self.head
        while cur.next:
            cur = cur.next
        cur.next = new

    def insert(self, val, ind):
        new = Node(val)

        # Insert at head
        if ind == 0:
            new.next = self.head
            self.head = new
            return

        cur = self.head
        for i in range(ind - 1):
            if not cur or not cur.next and i < ind - 2:
                raise IndexError("Index out of range")
            cur = cur.next

        new.next = cur.next
        cur.next = new
